### DISCUSSION TITLE
How to use circuit-relay to connect peers via webrtc

### DISCUSSION THREAD

**Discussion by sleep9** [libp2p/js-libp2p#2445] [Q&A]
Currently trying to connect alice (node.js or react-native - behind NAT) and bobs (node.js - public IP) via webrtc through a circuit relay server (node.js - public IP)

I cannot hardcode bobs multiaddr as many bobs may exist in my network. Only the circuit relays can be hardcoded.

Say I have a connection through a circuit relay with an example connection multiaddr like so `/ip4/${ip_address}/tcp/${port_number}/ws/p2p/12D3KooWSFr7xJnyREpieyX5sGw2mCzc4k6LEa2KEBkXNhYXoVhB/p2p-circuit/webrtc/p2p/12D3KooWDYhZLWM7XagHSf9HSZT338UN9LesS1ckQRhdnnQSSkk6`

Dialing the peer simply returns the same connection multiaddr type (p2p-circuit/webrtc)
`let res = await node.dial(connection?.remoteAddr);
`
I am expecting a webrtc connection so I can negotiate a webrtc protocol. Here is an example of what I am expecting to pass.

```
node.addEventListener("connection:open", async (event) => {
    
         const peerList = node.getConnections()
        .filter((connection) => {
            const protos: string[] = connection.remoteAddr.protoNames();
            if (protos.includes("p2p") && protos.includes("p2p-circuit") && protos.includes("webrtc")) {
              return true;
            }
            return false;
          })

          for(var j=0;j<peerList.length;j++) {
            
            let connection:Connection = peerList[j];
            let peer = connection?.remotePeer.toString();

            if(!hasPeer(peer)) {
                const raw_stream: Stream = await connection.newStream("/my_protocol/1.0.0"); <-- does not return
                const stream: LengthPrefixedStream = lpStream(raw_stream);
                await stream.write(Buffer.from('hello bob','utf8'));
                
            }
        }
}
```
**Edit**
The following works from node.js alice but not react-native alice. It seems as though webRTC is broken with React-native. 

```
let raw_stream = await node.dialProtocol(connection?.remoteAddr,'/my_protocol/1.0.0')
const stream: LengthPrefixedStream = lpStream(raw_stream);
await stream.write(Buffer.from('from client','utf8'));
```

**Comment by maschad** [comment]
Not all transports are fully compatible with react native at the moment, it may be more appropriate to raise this as an issue on the [react-native-repo](https://github.com/ipfs-shipyard/js-libp2p-react-native)

**Comment by sleep9** [comment]
Issue has been raised here https://github.com/ipfs-shipyard/js-libp2p-react-native/issues/7 with a bounty.
@libp2p/webrtc depends on react-native-webrtc, do you know which of these has the issue with unsupported transports?
